File: Leetcode Daily Challenge/026-August-2025/002-rearranging-fruits.cpp
// https://leetcode.com/problems/rearranging-fruits/




#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    typedef long long ll;

    // TC : O(N*log(N)) 
    // SC : O(N)
    long long minCost(vector<int>& basket1, vector<int>& basket2) {
        unordered_map<int, int> mpp;
        int minEle = INT_MAX;

        for (int &val : basket1) {
            mpp[val] ++;
            minEle = min(minEle, val);
        }    
        for (int &val : basket2) {
            mpp[val] --;
            minEle = min(minEle, val);
        }

        vector<int> store;  // stores only those elements which needs to be swapped
        for (auto &it : mpp) {
            int ele  = it.first;
            int freq = it.second;

            if (!freq) continue;

            if (freq % 2) return -1;

            for (int t = 1; t <= abs(freq) / 2; t ++) {
                store.push_back(ele);
            }
        }

        // sort(begin(store), end(store));
        
        // Using nth_element for partial sorting is more efficient than full sort (O(n log n)); 
        // it runs in linear time on average.
        nth_element(begin(store), begin(store) + store.size() / 2, end(store));
        
        ll result = 0;
        for (int i = 0; i < store.size() / 2; i ++) {
            result += min(store[i], minEle * 2);
        }
        return result;
    }
};

File: Leetcode Daily Challenge/026-August-2025/002-rearranging-fruits.java
// https://leetcode.com/problems/rearranging-fruits/




import java.util.*;

class Solution {
    // TC : O(N*log(N))
    // SC : O(N)
    public long minCost(int[] basket1, int[] basket2) {
        Map<Integer, Integer> mpp = new HashMap<>();
        int minEle = Integer.MAX_VALUE;

        for (int val : basket1) {
            mpp.put(val, mpp.getOrDefault(val, 0) + 1);
            minEle = Math.min(minEle, val);
        }
        for (int val : basket2) {
            mpp.put(val, mpp.getOrDefault(val, 0) - 1);
            minEle = Math.min(minEle, val);
        }

        List<Integer> finalList = new ArrayList<>();

        for (Map.Entry<Integer, Integer> entry : mpp.entrySet()) {
            Integer key = entry.getKey();
            Integer cnt = entry.getValue();

            if (cnt == 0) continue;
            if (cnt % 2 == 1) return -1;

            for (int t = 1; t <= Math.abs(cnt) / 2; t ++) {
                finalList.add(key);
            }
        }

        Collections.sort(finalList);

        long result = 0;
        for (int i = 0; i < finalList.size() / 2; i ++) {
            result += Math.min(finalList.get(i), minEle * 2);
        }
        return result;
    }
}